cardano-cli/test/cardano-cli-test/Test/Cli/Parser.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cli.Parser
+  ( hprop_integral_reader
+  , hprop_pair_integral_reader
+  )
+where
+
+import           Cardano.CLI.EraBased.Options.Common (integralParsecParser,
+                   pairIntegralParsecParser)
+
+import           Data.Bits (Bits)
+import           Data.Data (Typeable)
+import           Data.Either (isLeft, isRight)
+import           Data.Word (Word16)
+import qualified Text.Parsec as Parsec
+
+import           Hedgehog (Property, assert)
+import           Hedgehog.Extras (propertyOnce)
+
+-- | Execute me with:
+-- @cabal test cardano-cli-test --test-options '-p "/integral reader/"'@
+hprop_integral_reader :: Property
+hprop_integral_reader = propertyOnce $ do
+  assert $ isRight $ parse @Word "0"
+  assert $ isRight $ parse @Word "42"
+  assert $ isLeft $ parse @Word "-1"
+  assert $ isLeft $ parse @Word "18446744073709551616"
+  assert $ isLeft $ parse @Word "-1987090"
+
+  assert $ isRight $ parse @Word16 "0"
+  assert $ isRight $ parse @Word16 "42"
+  assert $ isLeft $ parse @Word16 "-1"
+  assert $ isLeft $ parse @Word16 "65536"
+  assert $ isLeft $ parse @Word16 "298709870987"
+  assert $ isLeft $ parse @Word16 "-1987090"
+ where
+  parse :: (Typeable a, Integral a, Bits a) => String -> Either String a
+  parse s =
+    case Parsec.runParser integralParsecParser () "" s of
+      Left parsecError -> Left $ show parsecError
+      Right x -> Right x
+
+-- | Execute me with:
+-- @cabal test cardano-cli-test --test-options '-p "/pair integral reader/"'@
+hprop_pair_integral_reader :: Property
+hprop_pair_integral_reader = propertyOnce $ do
+  assert $ isRight $ parse @Word "(0, 0)"
```

Comment:
We can improve these tests with hedgehog generators. Example:
```diff
diff --git a/cardano-cli/test/cardano-cli-test/Test/Cli/Parser.hs b/cardano-cli/test/cardano-cli-test/Test/Cli/Parser.hs
index 1bf19f8b8..c72c8a093 100644
--- a/cardano-cli/test/cardano-cli-test/Test/Cli/Parser.hs
+++ b/cardano-cli/test/cardano-cli-test/Test/Cli/Parser.hs
@@ -1,3 +1,5 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications #-}
 
 module Test.Cli.Parser
@@ -12,11 +14,14 @@ import           Cardano.CLI.EraBased.Options.Common (integralParsecParser,
 import           Data.Bits (Bits)
 import           Data.Data (Typeable)
 import           Data.Either (isLeft, isRight)
+import           Data.Proxy
 import           Data.Word (Word16)
 import qualified Text.Parsec as Parsec
 
-import           Hedgehog (Property, assert)
+import           Hedgehog
 import           Hedgehog.Extras (propertyOnce)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
 
 -- | Execute me with:
 -- @cabal test cardano-cli-test --test-options '-p "/integral reader/"'@
@@ -45,9 +50,8 @@ hprop_integral_reader = propertyOnce $ do
 -- @cabal test cardano-cli-test --test-options '-p "/pair integral reader/"'@
 hprop_pair_integral_reader :: Property
 hprop_pair_integral_reader = propertyOnce $ do
-  assert $ isRight $ parse @Word "(0, 0)"
-  assert $ isRight $ parse @Word "   (   0  ,   0   )"
-  assert $ isRight $ parse @Word " (18446744073709551615   ,    18446744073709551614   )   "
+  validArbitraryTuple <- forAll $ genNumberTuple (Proxy :: Proxy Word)
+  assert $ isRight $ parse @Word validArbitraryTuple
 
   assert $ isLeft $ parse @Word "(0, 0, 0)"
   assert $ isLeft $ parse @Word "(-1, 0)"
@@ -64,3 +68,18 @@ hprop_pair_integral_reader = propertyOnce $ do
     case Parsec.runParser pairIntegralParsecParser () "" s of
       Left parsecError -> Left $ show parsecError
       Right x -> Right x
+
+genNumberTuple
+  :: forall integral
+   . Integral integral
+  => Show integral
+  => Proxy integral -> Gen String
+genNumberTuple _ = do
+  x :: integral <- Gen.integral (Range.linear 0 100)
+  y :: integral <- Gen.integral (Range.linear 0 100)
+  space1 <- genArbitrarySpace
+  space2 <- genArbitrarySpace
+  return $ "(" ++ space2 ++ show x ++ space1 ++ "," ++ space2 ++ show y ++ space1 ++ ")"
+
+genArbitrarySpace :: Gen String
+genArbitrarySpace = Gen.string (Range.linear 0 5) (return ' ')

```

Reply:
We should as much as possible avoid unit tests.

Reply:
I took your suggestion and split the function into two, since the positive tests need to be run with `property`, while the negative tests need to run with `propertyOnce` :+1:

Reply:
> while the negative tests need to run with propertyOnce


Why? You could also write a generator to inject a random character (or multiple characters) between the brackets.

Reply:
Yes but I would need to rule out generators that would still yield a valid value (e.g. if the generated character is a number) and I believe writing a good generator for negative tests it overkill for testing this parser. So I'll stick to the manual negative tests to speed up merging.